sklearn/_loss/tests/test_loss.py

```diff
@@ -981,6 +982,44 @@ def test_binomial_and_multinomial_loss(global_random_seed):
     )
 
 
+def test_binomial_vs_alternative_formulation():
+    """Test that both formulations of the binomial deviance agree.
+
+    Often, the binomial deviance or log loss is written in terms of a variable
+    z in {-1, +1}, but we use y in {0, 1}, hence z = 2 * y - 1.
+    ESL II Eq. (10.18):
+
+        -loglike(z, f) = log(1 + exp(-2 * z * f))
+
+    Note:
+        - ESL 2*f = raw_prediction, hence the factor 2 of ESL disappears.
+        - Deviance = -2*loglike + .., but HalfBinomialLoss is half of the
+          deviance, hence the factor of 2 cancels in the comparison.
+    """
+
+    def alt_loss(y, raw_pred):
+        z = 2 * y - 1
+        return np.mean(np.log(1 + np.exp(-z * raw_pred)))
+
+    bin_loss = HalfBinomialLoss()
+
+    test_data = product(
+        (np.array([0.0, 0, 0]), np.array([1.0, 1, 1])),
+        (np.array([-5.0, -5, -5]), np.array([3.0, 3, 3])),
+    )
```

Comment:
The first `for` loop will consume this `product` and not allow for the second for loop to run. To fix this:

```suggestion
    test_data = list(
        product(
            (np.array([0.0, 0, 0]), np.array([1.0, 1, 1])),
            (np.array([-5.0, -5, -5]), np.array([3.0, 3, 3])),
        )
    )
```

With this change the `alt_gradient` check will run. (BTW the gradient check fails locally for me)

Reply:
Thanks for the fix. This means this test was wrong all time long...

Reply:
Would it not be preferable to parametrize the test directly? e.g.
```python
@pytest.mark.parametrize("y", (np.array([0.0, 0, 0]), np.array([1.0, 1, 1])))
@pytest.mark.parametrize("y_pred", (np.array([-5.0, -5, -5]), np.array([3.0, 3, 3])))
```

That way we could easily further parametrize dtype with
```python
@pytest.mark.parametrize("dtype", (np.float32, np.float64))
```
and eventually
```python
datum = (y.astype(dtype), y_pred.astype(dtype))
```

Also how much do we care for precision? `hypothesis` produces a few falsifying examples including 
```python
y_true = array([4.01900735e-14])
y_pred = array([-25.])
```
which results in

```python
E       AssertionError: assert 1.4892695702176324e-11 == 1.38880018595...e-11 ± 1.0e-12
E         comparison failed
E         Obtained: 1.4892695702176324e-11
E         Expected: 1.388800185954457e-11 ± 1.0e-12
```

Reply:
@Micky774 Do you want to take over? I won‘t have time the next month and my intention was to move this test here because it will be deleted in the GBT code with the replacement of the loss functions. That‘s why it‘s such old (not ideal) test code.

Reply:
Sure thing :)

Though I would like your opinion on the matter of precision for the falsifying example I proved, and whether that's something we can tolerate (I simply do not know).

Reply:
The formulation according to ESL only wirks for y_true -1 or +1. The binomial loss, otoh, can deal with all values in the interval [0,1]. That‘s why I prefer that one.
The failing example of hypothesis sets y_true to 1e-14 instead of 0.
I hope this solves the case.